File: api/income.py
import json
from urllib.request import Request, urlopen
from urllib.parse import urlencode

FINANCIAL_API_KEY = "YOUR_NEW_API_KEY"
FINANCIAL_BASE = "https://api.financialdata.net/v3"

def handler(request):
    query = {k: v[0] for k, v in request.get_query().items()} if request.get_query() else {}
    symbol = query.get("symbol")
    if not symbol:
        return {"statusCode": 400, "body": json.dumps({"error": "Missing symbol"})}
    
    params = urlencode({"period": "quarterly", "limit": 4, "token": FINANCIAL_API_KEY})
    url = f"{FINANCIAL_BASE}/financials/income-statement/{symbol.upper()}?{params}"
    try:
        with urlopen(Request(url), timeout=10) as resp:
            data = json.loads(resp.read().decode())
        if "financials" in data:
            quarters = [{"date": f.get("date"), "revenue": f.get("revenue"), "netIncome": f.get("netIncome")} for f in data["financials"]]
            return {"statusCode": 200, "body": json.dumps({"quarters": quarters}), "headers": {"Content-Type": "application/json"}}
        return {"statusCode": 404, "body": json.dumps({"error": "No income data"})}
    except Exception as e:
        return {"statusCode": 500, "body": json.dumps({"error": str(e)})}

File: api/income.js
export default async function handler(req, res) {
    const { symbol } = req.query;
    if (!symbol) return res.status(400).json({ error: 'Missing symbol' });
    const sym = symbol.toUpperCase();

    // FinancialData.net (quarterly)
    if (process.env.FINANCIAL_API_KEY) {
        try {
            const url = `https://api.financialdata.net/api/v1/income-statement/${sym}?period=quarterly&limit=4&token=${process.env.FINANCIAL_API_KEY}`;
            const data = await fetch(url).then(r => r.json());
            if (data?.financials?.length) {
                const quarters = data.financials.map(f => ({
                    date: f.date,
                    revenue: f.revenue,
                    netIncome: f.netIncome,
                }));
                return res.json({ quarters });
            }
        } catch (e) { console.warn('FinancialData income failed', e.message); }
    }

    // Fallback: FMP (annual)
    if (process.env.FMP_API_KEY) {
        try {
            const url = `https://financialmodelingprep.com/api/v3/income-statement/${sym}?limit=4&apikey=${process.env.FMP_API_KEY}`;
            const data = await fetch(url).then(r => r.json());
            if (data?.length) {
                const years = data.map(row => ({
                    date: row.date,
                    revenue: row.revenue,
                    netIncome: row.netIncome,
                }));
                return res.json({ quarters: years });
            }
        } catch (e) { console.warn('FMP income failed', e.message); }
    }

    return res.status(503).json({ error: 'No income data available' });
}
